### PR TITLE
[054]Add user to internship

### PR DESCRIPTION
**Pull request by ShirshovAlex**

### CODE REVIEW

File: src/test/java/by/exadel/internship/formTest/FormTest.java [hunk 1]
```diff
@@ -66,7 +66,8 @@ public void givenFormWithFile_checkTestData()
         assertEquals(formFullDTO.getEmail(), "string");
         assertEquals(formFullDTO.getEnglishLevel(), EnglishLevel.A0);
         assertEquals(formFullDTO.getExperience(), "string");
-        assertEquals(formFullDTO.getFilePath(), "text.txt");
+        //wrong test - (will learn and use @Mock @MocKBean in next PR)
+        assertEquals(formFullDTO.getFilePath().contains("https://firebasestorage.googleapis.com/v0/b/gs://internship-project-e202a.appspot.com/o/"), true);
```
Comment (mortal-gpt) [inline src/test/java/by/exadel/internship/formTest/FormTest.java]: I would disable this test at all not to mess storage with test files.

File: src/main/java/by/exadel/internship/controller/InternshipController.java [hunk 2]
```diff
@@ -71,4 +72,12 @@ public UserInternshipDTO restoreInternship(@PathVariable UUID internshipId) {
     public UserInternshipDTO saveInternship(@RequestBody UserInternshipDTO internshipDTO) {
         return internshipService.saveInternship(internshipDTO);
     }
+
+    @SuperAdminAccessControl
+    @PostMapping("/addUser")
+    @ApiOperation("Add User to Internship")
+    @ResponseStatus(HttpStatus.OK)
+    public boolean addUser(@RequestParam UUID userId, @RequestParam(name = "internshipId") UUID internshipId){
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/controller/InternshipController.java]: Why first @RequestParam doesn't have name attribute, but the second does? Bring parameters to the general view.

Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/controller/InternshipController.java]: Returning just boolean doesn't look like good solution

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 3]
```diff
@@ -23,11 +25,13 @@
 @Slf4j
 public class InternshipServiceImpl implements InternshipService {
 
+    private static final String SIMPLE_CLASS_NAME = InternshipService.class.getSimpleName();
     private final InternshipRepository internshipRepository;
+    private final UserRepository userRepository;
+    //    private final UserInternshipMapper userInternshipMapper;
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: Remove if you don't need them

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 4]
```diff
@@ -131,6 +135,18 @@ public UserInternshipDTO saveInternship(UserInternshipDTO internshipDTO) {
         return userInternshipDTO;
     }
 
+    public boolean addUser(UUID userId, UUID internshipId) {
+        internshipRepository.getOne(internshipId);
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: why are you getting it if not using later?

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 5]
```diff
@@ -131,6 +135,18 @@ public UserInternshipDTO saveInternship(UserInternshipDTO internshipDTO) {
...
+        User user = userRepository.getOne(userId);
+        for (Internship internship : user.getInternships()) {
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: I don't like this solution of checking if user is already assigned to the internship. Change please

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/test/java/by/exadel/internship/formTest/FormTest.java (hunk 1 above)
Comment (ShirshovAlex) [reply src/test/java/by/exadel/internship/formTest/FormTest.java]: done

File: src/main/java/by/exadel/internship/controller/InternshipController.java (hunk 2 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/controller/InternshipController.java]: done

Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/controller/InternshipController.java]: change to void

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 3 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 4 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: removed

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 5 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: done

File: src/main/java/by/exadel/internship/repository/InternshipRepository.java [hunk 6]
```diff
@@ -33,4 +34,13 @@
     @Modifying
     @Query(value = "UPDATE Internship i SET i.deleted=true WHERE i.id= :internshipId")
     void deleteById(@Param("internshipId") UUID internshipId);
+
+    @Transactional
```
Comment (OlgaEvdokimova) [inline src/main/java/by/exadel/internship/repository/InternshipRepository.java]: Is it nessesary to put the annotation @Transactional in the repository? Maybe it's better to put it in your Service?

File: src/main/java/by/exadel/internship/controller/InternshipController.java [hunk 7]
```diff
@@ -71,4 +72,12 @@ public UserInternshipDTO restoreInternship(@PathVariable UUID internshipId) {
     public UserInternshipDTO saveInternship(@RequestBody UserInternshipDTO internshipDTO) {
         return internshipService.saveInternship(internshipDTO);
     }
+
+    @SuperAdminAccessControl
+    @PostMapping("/addUser")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/controller/InternshipController.java]: I guess as `internshipId` is mandatory variable and it's InternshipController it'll be better to move it to path variable.

```suggestion
    @PostMapping("/{internshipId}/addUser")
```

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 8]
```diff
@@ -131,6 +132,20 @@ public UserInternshipDTO saveInternship(UserInternshipDTO internshipDTO) {
...
     }
 
+    public void addUser(UUID userId, UUID internshipId) {
+        userRepository.findById(userId)
+                .orElseThrow(() -> new NotFoundException("User Not Found"));
+
+        internshipRepository.findById(internshipId)
+                .orElseThrow(() -> new NotFoundException("Internship Not Found"));
+
+        int exist = internshipRepository.checkUserExists(userId, internshipId);
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: Why do you double check user?

File: src/main/java/by/exadel/internship/repository/InternshipRepository.java (hunk 6 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/repository/InternshipRepository.java]: done

File: src/main/java/by/exadel/internship/controller/InternshipController.java (hunk 7 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/controller/InternshipController.java]: done

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 8 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: 1st time we check user 
2nd time we check if our user already exist with our internship table

File: src/main/java/by/exadel/internship/controller/InternshipController.java [hunk 9]
```diff
@@ -71,4 +72,12 @@ public UserInternshipDTO restoreInternship(@PathVariable UUID internshipId) {
     public UserInternshipDTO saveInternship(@RequestBody UserInternshipDTO internshipDTO) {
         return internshipService.saveInternship(internshipDTO);
     }
+
+    @SuperAdminAccessControl
+    @PostMapping("/{internshipId}/addUser")
+    @ApiOperation("Add User to Internship")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/controller/InternshipController.java]: excessive
```suggestion
```

File: src/main/java/by/exadel/internship/repository/InternshipRepository.java [hunk 10]
```diff
@@ -33,4 +34,13 @@
     @Modifying
     @Query(value = "UPDATE Internship i SET i.deleted=true WHERE i.id= :internshipId")
     void deleteById(@Param("internshipId") UUID internshipId);
+
+//    @Transactional
+    @Modifying
+    @Query(value = "INSERT INTO user_internship(ui_u_id, ui_inship_id) VALUES(:user , :internship)", nativeQuery = true)
+    void addUserToInternship(@Param("user") UUID userId, @Param("internship") UUID internshipId);
+
+//    @Transactional
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/repository/InternshipRepository.java]: remove at all

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 8 above)
Comment (kzaretskaya) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: rename method `checkUserExists` to more appropriate name
For example, `checkUserAssign`

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 11]
```diff
@@ -131,6 +132,21 @@ public UserInternshipDTO saveInternship(UserInternshipDTO internshipDTO) {
...
+        if (exist > 0) {
+            throw new RuntimeException("ALREADY EXIST");
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: what `ALREADY EXIST`?) May be, `User is already assigned to Intership`.

File: src/main/java/by/exadel/internship/controller/InternshipController.java (hunk 9 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/controller/InternshipController.java]: done

File: src/main/java/by/exadel/internship/repository/InternshipRepository.java (hunk 10 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/repository/InternshipRepository.java]: done

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 11 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java (hunk 8 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java [hunk 12]
```diff
@@ -131,6 +132,21 @@ public UserInternshipDTO saveInternship(UserInternshipDTO internshipDTO) {
...
+    public void addUser(UUID userId, UUID internshipId) {
+        userRepository.findById(userId)
+                .orElseThrow(() -> new NotFoundException("User Not Found"));
+
+        internshipRepository.findById(internshipId)
+                .orElseThrow(() -> new NotFoundException("Internship Not Found"));
+
+        int exist = internshipRepository.checkUserAssign(userId, internshipId);
+        if (exist > 0) {
+            throw new RuntimeException("User is already assigned to Internship");
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/InternshipServiceImpl.java:145]: Let's do separate exception in another PR.

**Review by kzaretskaya — APPROVED** [review]

**Review by viktarkolbik — APPROVED** [review]

**Review by SergeyKoval — APPROVED** [review]
